internal/nodes/nodes.go

```diff
@@ -0,0 +1,139 @@
+/*
+SPDX-License-Identifier: Apache-2.0
+
+Copyright Contributors to the Submariner project.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package nodes
+
+import (
+	"context"
+	goerrors "errors"
+	"fmt"
+	"github.com/pkg/errors"
+	"github.com/submariner-io/submariner-operator/internal/constants"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	"strings"
+	"time"
+)
+
+// labels the specified worker node as a gateway node
+func LabelGateways(clientset kubernetes.Interface, gatewayNode struct{Node string}) error {
+
+	if gatewayNode.Node == "" {
+		fmt.Printf("no gateway nodes specified, selecting one of the worker node as gateway node")
+		workerNodes, err := GetAllWorkerNames(clientset)
+		if err != nil {
+			return err
+		}
+		gatewayNode.Node = workerNodes[0]
+	}
+	err := addLabels(clientset, gatewayNode.Node, map[string]string{constants.SubmarinerGatewayLabel: constants.TrueLabel})
+	return errors.Wrap(err, fmt.Sprintf("error labeling node %q as a gateway", gatewayNode.Node))
+}
+
+// GetAllWorkerNames returns all worker nodes
+func GetAllWorkerNames(clientset kubernetes.Interface) ([]string, error) {
+	workerNodes, err := clientset.CoreV1().Nodes().List(
+		context.TODO(), metav1.ListOptions{LabelSelector: "node-role.kubernetes.io/worker"})
+	if err != nil {
+		return nil, errors.Wrap(err, "error listing Nodes")
+	}
+
+	if len(workerNodes.Items) == 0 {
+		// In some deployments (like KIND), worker nodes are not explicitly labelled. So list non-master nodes.
+		workerNodes, err = clientset.CoreV1().Nodes().List(
+			context.TODO(), metav1.ListOptions{LabelSelector: "!node-role.kubernetes.io/master"})
+		if err != nil {
+			return nil, errors.Wrap(err, "error listing Nodes")
+		}
+
+		if len(workerNodes.Items) == 0 {
+			return nil, fmt.Errorf("* No worker node found to label as the gateway\n")
+		}
+	}
+
+	allNodeNames := []string{}
+	for i := range workerNodes.Items {
+		allNodeNames = append(allNodeNames, workerNodes.Items[i].GetName())
+	}
+	return allNodeNames, nil
+}
+
+// ListGateways returns all nodes labelled as gateway
+func ListGateways(clientset kubernetes.Interface) (bool, error) {
```

Comment:
Retrying this comment as it got removed by the last push...

This should take a `kubernetes.Interface` instead of `rest.Config`.

However it's advertised to list and return all nodes labeled as gateway but it actually returns whether or not there are any gateways. I think it should do as advertised. Also it shouldn't print out the gateways as a side-effect.